### PR TITLE
add notification task

### PR DESCRIPTION
**Pull request by mdayican**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/BPS-1085

### Change description ###

add spring schedule to handle notification messages

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    public NotificationMessageProcessTask(
+        NotificationMessageProcessor notificationMessageProcessor
+    ) {
+        this.notificationMessageProcessor = notificationMessageProcessor;
+    }
+
+    @Scheduled(fixedDelayString = "${scheduling.task.notifications-consume.delay}")
+    public void consumeMessages() {
+        log.info("Started the job consuming notification messages");
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]: i think we usually do `"Started/Finished {task-name} task"` where `{task-name}` is statically defined string with hyphen separated words if necessary, e.g. `consume-notifications` in this case

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        } catch (InterruptedException exception) {
+            logTaskError(exception);
+            Thread.currentThread().interrupt();
+        } catch (Exception exception) {
+            logTaskError(exception);
+        }
+    }
+
+    private void logTaskError(Exception exception) {
+        log.error("An error occurred when running the 'consume notification messages' task", exception);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]: This message should also be reflected by revised task start/finish log entries

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+            do {
+                queueMayHaveMessages = notificationMessageProcessor.processNextMessage();
+            } while (queueMayHaveMessages);
+
+            log.info("Finished the job consuming notification messages");
+        } catch (InterruptedException exception) {
+            logTaskError(exception);
+            Thread.currentThread().interrupt();
+        } catch (Exception exception) {
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]: do we want to be vague here? i believe it's `ServiceBusException` and if any other there is/will be - it'll be handled with [Scheduler Config](https://github.com/hmcts/reform-scan-notification-service/blob/master/src/main/java/uk/gov/hmcts/reform/notificationservice/config/SchedulerConfiguration.java#L46-L49) which then can be addressed appropriately

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]: change the type servicebusexception

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import uk.gov.hmcts.reform.notificationservice.service.NotificationMessageProcessor;
+
+@Service
+@ConditionalOnProperty(value = "scheduling.task.notifications-consume.enabled", matchIfMissing = true)
+public class NotificationMessageProcessTask {
+
+    private static final Logger log = LoggerFactory.getLogger(NotificationMessageProcessTask.class);
+
+    private final NotificationMessageProcessor notificationMessageProcessor;
+    private static final String TASK_NAME = "notification consume task";
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]: Still wrong task name ;)

```suggestion
    private static final String TASK_NAME = "consume-notifications";
```

this ^ is just a suggestion but wording sounds good to me :man_shrugging:

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    public NotificationMessageProcessTask(
+        NotificationMessageProcessor notificationMessageProcessor
+    ) {
+        this.notificationMessageProcessor = notificationMessageProcessor;
+    }
+
+    @Scheduled(fixedDelayString = "${scheduling.task.notifications-consume.delay}")
+    public void consumeMessages() {
+        log.info("Start {}", TASK_NAME);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]:
```suggestion
        log.info("Started {} task", TASK_NAME);
```

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        try {
+            boolean queueMayHaveMessages;
+
+            do {
+                queueMayHaveMessages = notificationMessageProcessor.processNextMessage();
+            } while (queueMayHaveMessages);
+
+            log.info("Finish {}", TASK_NAME);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]:
```suggestion
            log.info("Finished {} task", TASK_NAME);
```

File: src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        } catch (InterruptedException exception) {
+            logTaskError(exception);
+            Thread.currentThread().interrupt();
+        } catch (ServiceBusException exception) {
+            logTaskError(exception);
+        }
+    }
+
+    private void logTaskError(Exception exception) {
+        log.error("Error occurred in {}", TASK_NAME, exception);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/notificationservice/task/NotificationMessageProcessTask.java]:
```suggestion
        log.error("Error occurred during {} task", TASK_NAME, exception);
```

**Pull request merged**